// office-ui-fabric-react/lib-amd/components/Button/PrimaryButton/PrimaryButton.styles.js
define(["require", "exports", "../../../Styling", "../../../Utilities", "../DefaultButton/DefaultButton.styles", "../SplitButton/SplitButton.styles"], function (require, exports, Styling_1, Utilities_1, DefaultButton_styles_1, SplitButton_styles_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var MS_HIGHCONTRAST_ACTIVE = '@media screen and (-ms-high-contrast: active)';
    exports.getStyles = Utilities_1.memoizeFunction(function (theme, customStyles) {
        var palette = theme.palette;
        var splitButtonStyles = SplitButton_styles_1.getStyles(theme);
        var defaultButtonStyles = DefaultButton_styles_1.getStyles(theme, customStyles, '0px', palette.white);
        var primaryButtonStyles = {
            root: (_a = {
                    backgroundColor: palette.themePrimary,
                    color: palette.white
                },
                _a[MS_HIGHCONTRAST_ACTIVE] = {
                    borderColor: 'Highlight',
                    color: 'Highlight'
                },
                _a),
            rootHovered: {
                backgroundColor: palette.themeDark,
                color: palette.white
            },
            rootPressed: {
                backgroundColor: palette.themePrimary,
                color: palette.white
            },
            rootExpanded: {
                backgroundColor: palette.themePrimary,
                color: palette.white
            },
            rootChecked: {
                backgroundColor: palette.themeDark,
                color: palette.white,
            },
            rootCheckedHovered: {
                backgroundColor: theme.palette.neutralLight,
                color: theme.palette.black
            }
        };
        return Styling_1.mergeStyleSets(defaultButtonStyles, primaryButtonStyles, splitButtonStyles, customStyles);
        var _a;
    });
});
