// @microsoft/api-extractor/lib/ast/AstMethod.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var AstItem_1 = require("./AstItem");
var AstMember_1 = require("./AstMember");
var AstParameter_1 = require("./AstParameter");
var TypeScriptHelpers_1 = require("../TypeScriptHelpers");
var ApiDefinitionReference_1 = require("../ApiDefinitionReference");
/**
 * This class is part of the AstItem abstract syntax tree. It represents functions that are members of
 * classes, interfaces, or nested type literal expressions. Unlike AstFunctions, AstMethods can have
 * access modifiers (public, private, etc.) or be optional, because they are members of a structured type
 *
 * @see AstFunction for functions that are defined inside of a package
 */
var AstMethod = (function (_super) {
    __extends(AstMethod, _super);
    function AstMethod(options) {
        var _this = _super.call(this, options) || this;
        _this.kind = AstItem_1.AstItemKind.Method;
        var methodDeclaration = options.declaration;
        // Parameters
        if (methodDeclaration.parameters) {
            _this.params = [];
            for (var _i = 0, _a = methodDeclaration.parameters; _i < _a.length; _i++) {
                var param = _a[_i];
                var declarationSymbol = TypeScriptHelpers_1.default.tryGetSymbolForDeclaration(param);
                var astParameter = new AstParameter_1.default({
                    extractor: _this.extractor,
                    declaration: param,
                    declarationSymbol: declarationSymbol,
                    jsdocNode: param
                });
                _this.innerItems.push(astParameter);
                _this.params.push(astParameter);
            }
        }
        // tslint:disable-next-line:no-bitwise
        _this._isConstructor = (options.declarationSymbol.flags & ts.SymbolFlags.Constructor) !== 0;
        // Return type
        if (!_this.isConstructor) {
            if (methodDeclaration.type) {
                _this.returnType = methodDeclaration.type.getText();
            }
            else {
                _this.returnType = 'any';
                _this.hasIncompleteTypes = true;
            }
        }
        return _this;
    }
    Object.defineProperty(AstMethod.prototype, "isConstructor", {
        /**
         * Returns true if this member represents a class constructor.
         */
        get: function () {
            return this._isConstructor;
        },
        enumerable: true,
        configurable: true
    });
    AstMethod.prototype.onCompleteInitialization = function () {
        _super.prototype.onCompleteInitialization.call(this);
        // If this is a class constructor, and if the documentation summary was omitted, then
        // we fill in a default summary versus flagging it as "undocumented".
        // Generally class constructors have uninteresting documentation.
        if (this.isConstructor) {
            if (this.documentation.summary.length === 0) {
                this.documentation.summary.push({
                    kind: 'textDocElement',
                    value: 'Constructs a new instance of the '
                });
                var scopedPackageName = ApiDefinitionReference_1.default
                    .parseScopedPackageName(this.extractor.package.name);
                this.documentation.summary.push({
                    kind: 'linkDocElement',
                    referenceType: 'code',
                    scopeName: scopedPackageName.scope,
                    packageName: scopedPackageName.package,
                    exportName: this.parentContainer.name,
                    value: this.parentContainer.name
                });
                this.documentation.summary.push({
                    kind: 'textDocElement',
                    value: ' class'
                });
            }
            this.needsDocumentation = false;
        }
    };
    return AstMethod;
}(AstMember_1.default));
exports.default = AstMethod;
